### PR TITLE
Add health feature to splinterD

### PR DESCRIPTION
**Pull request by RyanLassigBanks**
Adds Health service skeleton, connects it to the SplinterD, and gates it behind the `health` feature.

### CODE REVIEW

File: cli/src/main.rs [hunk 1]
```diff
@@ -62,6 +62,13 @@ fn run() -> Result<(), CliError> {
                 (@arg quiet: -q --quiet "do not display output")
             )
         )
+        (@subcommand health =>
+            (about: "Splinter commands for inspecting network integrety and health")
```
Comment (peterschwarz) [inline cli/src/main.rs]: Should match the starting case of the other descriptions - same for the `status` subcommand.

File: splinterd/src/routes/status.rs [hunk 2]
```diff
@@ -12,24 +12,35 @@
...
 
 pub fn get_status(
     node_id: String,
     endpoint: String,
+    peer_map: Arc<RwLock<PeerMap>>,
 ) -> Box<dyn Future<Item = HttpResponse, Error = Error>> {
+    let connected_peers = if let Ok(inner) = peer_map.read() {
+        inner.peer_endpoints()
+    } else {
+        return Box::new(HttpResponse::InternalServerError().finish().into_future());
```
Comment (peterschwarz) [inline splinterd/src/routes/status.rs]: Can we log the error here?  Otherwise, we won't know about the internal error.

File: cli/src/main.rs [hunk 3]
```diff
@@ -62,6 +62,13 @@ fn run() -> Result<(), CliError> {
...
+            (@subcommand status =>
+                (about: "Check status of node")
```
Comment (chenette) [inline cli/src/main.rs]: The description seems too terse. How about clarifying the specific information that `status` displays?

For example:
"shows a node's connected peers, {status?} endpoint, node ID, and version {of what?}"

Or, with more details:  
"queries the status endpoint for a node's connected peers, node ID, and other status information"

_Style note: The first word should start with a lower-case letter and end with "s" to be consistent with the other descriptions (keygen & keyregistry say "generates")._

File: cli/src/main.rs [hunk 4]
```diff
@@ -62,6 +62,14 @@ fn run() -> Result<(), CliError> {
                 (@arg quiet: -q --quiet "do not display output")
             )
         )
+        (@subcommand health =>
+            (about: "displays information about network health")
```
Comment (vaporos) [inline cli/src/main.rs]: The first letter should be capitalized so output in "carto run -- --help" is consistent.

Comment (vaporos) [reply cli/src/main.rs]: This description probably becomes outdated when we add the active tests. Maybe "Health commands" to match "Administrative commands", though something slightly more descriptive (but not verbose) might be better.

File: cli/src/main.rs [hunk 5]
```diff
@@ -62,6 +62,14 @@ fn run() -> Result<(), CliError> {
...
+            (@subcommand status =>
+                (about: "displays a node's version, endpoint, node id, and a list \
```
Comment (vaporos) [inline cli/src/main.rs]: Shorten this so it fits in <80 columns. This isn't always possible to achieve w/clap but where we can, we should. Maybe "Display node-level health information".

File: cli/src/main.rs [hunk 6]
```diff
@@ -62,6 +62,14 @@ fn run() -> Result<(), CliError> {
...
+                of the endpoints of its connected peers")
+                (@arg url: -u +takes_value "URL of node")
```
Comment (vaporos) [inline cli/src/main.rs]: Add the long argument '--url'. Not sure if this should be -U or -u. I seem to recall -U being more popular for this but haven't found specific examples yet.

Comment (vaporos) [reply cli/src/main.rs]: Should we say "URL of node's health REST API" to clarify which "URL"?

File: cli/src/main.rs (hunk 4 above)
Comment (peterschwarz) [reply cli/src/main.rs]: If that's the case, it should be done in an additional commit, as the other messages are lowercase

**Comment by vaporos** [comment]
The status information should differentiate between "peers" (other splinter nodes) and connected services.

{
  "peers": [ {
    "endpoint": "tls://172.22.0.8:35612",
    "status": "connected"
  },
  {
 "endpoint": "tls://172.22.0.9:35612",
    "status": "disconnected",
    "error": "Connection refused",
  }],
  "service_processors": [
      {
        "endpoint": "inproc://admin-service"
        ...
      }
    ]
  ],
  "endpoint": "0.0.0.0:8043",
  "node_id": "bubba-node-000",
  "version": "0.3.3"
}

We should also add in a list of circuits and their health status:

"circuits": [
{
    "name": "foo",
    "members": [
        "tls://172.22.0.8:35612",
        "tls://172.22.0.9:35612"
    ],
    "peering_status": "incomplete"
}]

File: cli/src/action/mod.rs [hunk 7]
```diff
@@ -12,6 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 pub mod admin;
+pub mod health;
```
Comment (vaporos) [inline cli/src/action/mod.rs]: This new part of the CLI should be gated by a feature.

File: cli/Cargo.toml [hunk 8]
```diff
@@ -26,7 +26,9 @@ protobuf = "2"
 simple_logger = "0.4.0"
 libc = "0.2"
 splinter = { path = "../libsplinter" }
+reqwest = "0.9"
```
Comment (vaporos) [inline cli/Cargo.toml]: These should probably be pulled in by the health feature, when enabled.

File: libsplinter/src/network/mod.rs [hunk 9]
```diff
@@ -51,7 +51,7 @@ impl NetworkMessage {
     }
 }
 
-struct PeerMap {
+pub struct PeerMap {
```
Comment (vaporos) [inline libsplinter/src/network/mod.rs]: This feels like it should remain an implementation detail of the network layer, and if we expose it, we should consider if this is truly the right data structure across the API boundary. In any case, this does not differentiate between the types of "peers" (splinter<->splinter) vs (splinter<->service).

File: cli/src/main.rs (hunk 5 above)
Comment (chenette) [reply cli/src/main.rs]: Isn't there a way to provide multi-line documentation without violating clap's line-length restrictions? I am in favor of clear & helpful information over short lines.

File: splinterd/Cargo.toml [hunk 10]
```diff
@@ -43,6 +44,7 @@ tempdir = "0.3"
 
 [features]
 default = []
+health-service-enabled = [ "health" ]
```
Comment (peterschwarz) [inline splinterd/Cargo.toml]: this feature should probably be just `health-service` (as enabled state is determined by if the feature is included in the build)

File: splinterd/src/daemon.rs [hunk 11]
```diff
@@ -358,19 +370,43 @@ impl SplinterDaemon {
         })
         .expect("Error setting Ctrl-C handler");
 
-        main_loop_join_handle
-            .join()
-            .map_err(|_| StartError::ThreadError("Unable to join main loop".into()))?;
+        #[cfg(feature = "health")]
```
Comment (peterschwarz) [inline splinterd/src/daemon.rs]: Does this block need cover all the join logic?

File: splinterd/src/daemon.rs [hunk 12]
```diff
@@ -475,6 +511,70 @@ impl SplinterDaemon {
...
+#[cfg(feature = "health")]
+fn start_health_service(
+    mut transport: InprocTransport,
+    health_service: HealthService,
+    running: Arc<AtomicBool>,
+) -> Result<service::JoinHandles<Result<(), service::error::ServiceProcessorError>>, StartError> {
+    let start_health_service: std::thread::JoinHandle<
+        Result<service::JoinHandles<Result<(), service::error::ServiceProcessorError>>, StartError>,
+    > = thread::spawn(move || {
+        let incoming_capacity = 8;
```
Comment (peterschwarz) [inline splinterd/src/daemon.rs]: might as well make these constants, like the admin service equivalents.

File: splinterd/Cargo.toml (hunk 10 above)
Comment (vaporos) [reply splinterd/Cargo.toml]: Also, move it below experimental below and add to the experimental block.

**Pull request merged**